File: viewer/utils/pipeline/commands.py
"""
Pipeline command generation utilities.
"""
import json
import os
import subprocess
import logging
import time
from pathlib import Path
from datetime import datetime
from .config import get_reference_name, get_chemistry
from viewer.models import Main

# Set up logging
logger = logging.getLogger(__name__)

def create_bash_script(commands, script_name='temp_script.sh'):
    """Create a temporary bash script with the given commands"""
    script_path = Path(os.path.join('/tmp', script_name))
    
    with open(script_path, 'w') as f:
        f.write('#!/bin/bash\n')
        f.write('source /home/svc_bicore/genomics-cloud-services/gcs-cli/.venv/bin/activate\n')
        f.write('export AWS_PROFILE=aibs-bicore\n\n')
        
        if isinstance(commands, list):
            for cmd in commands:
                f.write(f'{cmd}\n')
        else:
            f.write(f'{commands}\n')
    
    # Make script executable
    os.chmod(script_path, 0o755)
    
    return script_path

def run_bash_script(script_path):
    """Run a bash script and return its output"""
    try:
        # Read script content to determine if we should log or execute
        with open(script_path, 'r') as script_file:
            script_content = script_file.read()
        
        # Check if this is a submission command (that should be logged instead of executed)
        is_submission_command = False
        if 'align tenx-' in script_content:
            is_submission_command = True
        
        # For submission commands, log instead of executing
        if is_submission_command:
            # Create logs directory if it doesn't exist
            log_dir = Path('pipeline_command_logs')
            log_dir.mkdir(exist_ok=True)
            
            # Create log file with timestamp
            timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
            log_path = log_dir / f'command_log_{timestamp}.txt'
            
            with open(log_path, 'w') as log_file:
                log_file.write(f"--- COMMAND LOG AT {datetime.now().isoformat()} ---\n")
                log_file.write(f"Script path: {script_path}\n\n")
                log_file.write("Script content:\n")
                log_file.write(script_content)
                log_file.write("\n--- END OF COMMAND LOG ---\n")
            
            # Mock a successful submission response
            return json.dumps({
                "demand_status": "SUBMITTED", 
                "demand_id": f"test-demand-{timestamp}"
            })
        
        # For non-submission commands (like job tracking), actually execute
        output = subprocess.check_output([script_path], stderr=subprocess.STDOUT, universal_newlines=True)
        return output
        
    except subprocess.CalledProcessError as e:
        logger.error(f"Script execution failed: {e.output}")
        return e.output
    except Exception as e:
        logger.error(f"Error with script: {str(e)}")
        return f"Error: {str(e)}"

def is_ingest_complete(fastq_name):
    """Check if ingest is complete for a given fastq name"""
    try:
        sample = Main.objects.get(fastq_name=fastq_name)
        return sample.ingest_status == 'Completed'
    except Main.DoesNotExist:
        return False

def determine_workflow(batch_name_from_vendor):
    """Determine workflow based on batch name from vendor"""
    if not batch_name_from_vendor:
        return None
    
    # Extract prefix (e.g., MTX from MTX-22030)
    parts = batch_name_from_vendor.split('-')
    if not parts:
        return None
    
    prefix = parts[0].upper()
    
    if prefix == 'MTX':
        return 'mtx'
    elif prefix == 'RTX':
        return 'rtx'
    else:
        return None

def create_mtx_alignment_command(sample):
    """Create MTX alignment command for a sample"""
    reference = get_reference_name(sample.get('organism_common_name', ''))
    load_name = sample.get('load_name', '')
    
    command = f'ocs fastqs align tenx-arc --reference-names "{reference}" --asset-name cellranger-arc --load-names "{load_name}" --notify-on FAILED --notify beagan.nguy@alleninstitute.org'
    
    return command

def create_rtx_alignment_command(sample):
    """Create RTX alignment command for a sample"""
    reference = get_reference_name(sample.get('organism_common_name', ''))
    load_name = sample.get('load_name', '')
    lib_prep = sample.get('library_prep', '')
    chemistry = get_chemistry(lib_prep)
    
    # Determine command based on library prep method
    if lib_prep in ['10xV3.1D', '10xRseq_Mult_noATAC', '10xV3.1_HT', '10Xv3.1']:
        command = f'ocs fastqs align tenx-rnaseq --reference-names "{reference}" --asset-name cellranger-rnaseq --load-names "{load_name}" --cellranger-addopts "--chemistry {chemistry} --include-introns"'
    elif lib_prep in ['10xV3.1_HT_CP', '10xV3.1_HT_CP-BC']:
        command = f'ocs fastqs align tenx-rnaseq-multi --asset-name cellranger-multi --reference-names "{reference}" --cellranger-addopts "--include-introns" --execution-priority HIGH --load-names "{load_name}"'
    elif lib_prep == '10xV4':
        command = f'ocs fastqs align tenx-rnaseq --reference-names "{reference}" --asset-name cellranger-rnaseq --load-names "{load_name}" --asset-tag 8.0.1 --cellranger-addopts "--chemistry {chemistry}"'
    else:
        logger.warning(f"Unknown library prep method: {lib_prep}, using default command")
        command = f'ocs fastqs align tenx-rnaseq --reference-names "{reference}" --asset-name cellranger-rnaseq --load-names "{load_name}" --cellranger-addopts "--chemistry {chemistry} --include-introns"'
    
    return command

def submit_sample_for_alignment(sample):
    """Submit a sample for alignment and return the result"""
    # Check if ingest is complete
    fastq_name = sample.get('fastq_name')
    if not is_ingest_complete(fastq_name):
        return {
            'status': 'error',
            'message': f'Ingest not complete for {fastq_name}',
            'fastq_name': fastq_name
        }
    
    # Get workflow from batch name
    batch_name = sample.get('batch_name_from_vendor', '')
    workflow = determine_workflow(batch_name)
    
    if not workflow:
        return {
            'status': 'error',
            'message': f'Could not determine workflow for {fastq_name} with batch name {batch_name}',
            'fastq_name': fastq_name
        }
    
    # Create alignment command based on workflow
    if workflow == 'mtx':
        command = create_mtx_alignment_command(sample)
    else:  # rtx
        command = create_rtx_alignment_command(sample)
    
    # Create and run bash script
    script_path = create_bash_script(command, f'submit_{fastq_name}.sh')
    output = run_bash_script(script_path)
    
    # Parse result
    try:
        # In test mode, output is already a JSON string
        result = json.loads(output)
        
        if 'demand_status' in result and result['demand_status'] == 'SUBMITTED':
            # Success
            return {
                'status': 'success',
                'message': f'Alignment submitted successfully for {fastq_name}',
                'fastq_name': fastq_name,
                'demand_id': result.get('demand_id', 'unknown'),
                'command': command
            }
        else:
            # Something went wrong
            return {
                'status': 'error',
                'message': f'Submission failed for {fastq_name}: {output}',
                'fastq_name': fastq_name
            }
    except json.JSONDecodeError:
        logger.error(f"Failed to parse submission output: {output}")
        # For test mode fallback
        if "Command logged" in output:
            return {
                'status': 'success',
                'message': f'Alignment command logged for {fastq_name}',
                'fastq_name': fastq_name,
                'demand_id': f'test-demand-{int(time.time())}',
                'command': command
            }
        return {
            'status': 'error',
            'message': f'Invalid response from OCS: {output}',
            'fastq_name': fastq_name
        } 

File: viewer/static/viewer/js/pipeline/commands.js
/**
 * commands.js
 * Handles generation of pipeline commands for different workflows
 */

class PipelineCommands {
    /**
     * Generate alignment command for a sample
     * @param {Object} sample - Sample data
     * @returns {string} Command string
     */
    static generateAlignmentCommand(sample) {
        const workflow = sample.workflow || this.determineWorkflow(sample);
        const fastqName = sample.fastq_name || sample.fastqName;
        const loadName = sample.load_name || '';
        const organism = sample.organism_common_name || '';
        const reference = this.getReference(organism);
        const libraryPrep = sample.library_prep || '';

        // Generate command according to workflow type
        if (workflow === 'MTX') {
            return `ocs fastqs align tenx-arc --reference-names "${reference}" --asset-name cellranger-arc --load-names "${loadName}" --notify-on FAILED --notify beagan.nguy@alleninstitute.org`;
        } else {
            // RTX workflow
            const chemistry = this.getChemistry(libraryPrep);

            // Determine command based on library prep method
            if (['10xV3.1D', '10xRseq_Mult_noATAC', '10xV3.1_HT', '10Xv3.1'].includes(libraryPrep)) {
                return `ocs fastqs align tenx-rnaseq --reference-names "${reference}" --asset-name cellranger-rnaseq --load-names "${loadName}" --cellranger-addopts "--chemistry ${chemistry} --include-introns"`;
            } else if (['10xV3.1_HT_CP', '10xV3.1_HT_CP-BC'].includes(libraryPrep)) {
                return `ocs fastqs align tenx-rnaseq-multi --asset-name cellranger-multi --reference-names "${reference}" --cellranger-addopts "--include-introns" --execution-priority HIGH --load-names "${loadName}"`;
            } else if (libraryPrep === '10xV4') {
                return `ocs fastqs align tenx-rnaseq --reference-names "${reference}" --asset-name cellranger-rnaseq --load-names "${loadName}" --asset-tag 8.0.1 --cellranger-addopts "--chemistry ${chemistry}"`;
            } else {
                return `ocs fastqs align tenx-rnaseq --reference-names "${reference}" --asset-name cellranger-rnaseq --load-names "${loadName}" --cellranger-addopts "--chemistry ${chemistry} --include-introns"`;
            }
        }
    }

    /**
     * Generate post-QC command for a sample
     * @param {Object} sample - Sample data
     * @returns {string} Command string
     */
    static generatePostQCCommand(sample) {
        const workflow = sample.workflow || this.determineWorkflow(sample);
        const fastqName = sample.fastq_name || sample.fastqName;

        if (workflow === 'MTX') {
            return `python run_mtx_postqc.py --config=pipeline_config.yaml --fastq=${fastqName}`;
        } else {
            return `python run_rtx_postqc.py --config=pipeline_config.yaml --fastq=${fastqName}`;
        }
    }

    /**
     * Determine workflow based on batch name
     * @param {Object} sample - Sample data
     * @returns {string} Workflow type (MTX or RTX)
     */
    static determineWorkflow(sample) {
        // Get batch name from vendor, default to empty string if not present
        const batchName = (sample.batch_name_from_vendor || '').toUpperCase();

        // Check for MTX workflow
        if (batchName.startsWith('MTX') || batchName.includes('ATX')) {
            return 'MTX';
        }

        // Check for RTX workflow
        if (batchName.startsWith('RTX') || !batchName.includes('-')) {
            return 'RTX';
        }

        // Default to RTX for unrecognized patterns
        return 'RTX';
    }

    /**
     * Get reference genome name for an organism
     * @param {string} organism - Organism common name
     * @returns {string} Reference genome name
     */
    static getReference(organism) {
        // Map organism to reference genome
        switch ((organism || '').toLowerCase()) {
            case 'human':
                return 'human_10x_grch38_genome_star2.7.1a';
            case 'mouse':
                return 'mouse_10x_mm10_genome_star2.7.1a';
            case 'rat':
                return 'rat_10x_rn6';
            case 'monkey':
            case 'macaque':
                return 'macaque_10x_mmul10';
            case 'armadillo':
                return 'armadillo_ncbi_mdasnov1-hap2_genome_star2-7-1a';
            default:
                return 'human_10x_grch38_genome_star2.7.1a'; // Default to human
        }
    }

    /**
     * Get chemistry value for a library prep method
     * @param {string} libraryPrep - Library prep method
     * @returns {string} Chemistry value
     */
    static getChemistry(libraryPrep) {
        // Map library prep to chemistry
        switch (libraryPrep) {
            case '10xV3.1D':
                return 'SC3Pv3';
            case '10xRseq_Mult_noATAC':
                return 'ARC-v1';
            case '10xV3.1_HT':
                return 'SC3Pv3HT';
            case '10xV4':
                return 'SC3Pv4';
            case '10Xv2':
                return 'SC3Pv2';
            default:
                return 'SC3Pv3'; // Default to SC3Pv3
        }
    }
}

// Expose the commands module
window.PipelineCommands = PipelineCommands; 